### PR TITLE
Added models for CourseFacilitator, TeacherProgressReport, and

### PR DESCRIPTION
**Pull request by aoby**
WorkshopOrganizerReport plus associated tests.

Added some extra validations and tests for existing pd models.

For report context, here's the [spec](https://docs.google.com/document/d/1DTu72puX8jAh1LOxGyRjDAc7wL-2P5mBZpm7GmPi15Y).

cc @joshlory.

### CODE REVIEW

File: dashboard/app/models/pd/workshop.rb [hunk 1]
```diff
@@ -35,16 +35,39 @@ class Pd::Workshop < ActiveRecord::Base
     COURSE_ECS = 'ECS',
     COURSE_CS_IN_A = 'CSinA',
     COURSE_CS_IN_S = 'CSinS',
-    COURSE_CSD = 'CSD'
+    COURSE_CSD = 'CSD',
+    COURSE_COUNSELOR_ADMIN = 'Counselor/Admin'
   ]
 
   STATE_NOT_STARTED = 'Not Started'
```
Comment (mehalshah) [inline dashboard/app/models/pd/workshop.rb]: Might be nicer to define this as

STATES = [
NOT_STARTED,
IN_PROGRESS,
ENDED
] for better readability

File: dashboard/app/models/pd/teacher_progress_report.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
+class Pd::TeacherProgressReport
+
+  def self.generate_report_for_user(user)
+    teachers =
+      if user.admin?
+        # All teachers in all workshops
+        Pd::Attendance.distinct_teachers
```
Comment (mehalshah) [inline dashboard/app/models/pd/teacher_progress_report.rb:6]: This is potentially very large. Does this need to be paginated?

File: dashboard/app/models/pd/teacher_progress_report.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      elsif user.district_contact?
+        # All teachers attending workshops, from districts for which I am a contact
+        Pd::Attendance.for_district(user.district_as_contact).distinct_teachers
+      elsif user.workshop_organizer?
+        # All teachers in workshops I organized
+        Pd::Workshop.in_state(Pd::Workshop::STATE_ENDED).organized_by(user).map do |workshop|
+          Pd::Attendance.for_workshop(workshop).distinct_teachers
+        end.flatten
+      else
+        []
```
Comment (mehalshah) [inline dashboard/app/models/pd/teacher_progress_report.rb]: If you get here then something is wrong with the permissions model. You should redirect to forbidden

File: dashboard/app/models/pd/workshop.rb (hunk 1 above)
Comment (aoby) [inline dashboard/app/models/pd/workshop.rb]: good suggestion. will update.

File: dashboard/app/models/pd/teacher_progress_report.rb (hunk 2 above)
Comment (aoby) [inline dashboard/app/models/pd/teacher_progress_report.rb:6]: That's a lower priority work item, for after facilitator summit. We plan to add a date filter to all the reports and probably pagination as well.

File: dashboard/app/models/pd/teacher_progress_report.rb (hunk 3 above)
Comment (aoby) [inline dashboard/app/models/pd/teacher_progress_report.rb]: good call. will update.

File: dashboard/app/models/pd/workshop_organizer_report.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    if workshop.course == Pd::Workshop::COURSE_CSF &&
+      [Pd::Workshop::TYPE_PUBLIC, Pd::Workshop::TYPE_PRIVATE].include?(workshop.workshop_type)
+
+      qualified = true
+      num_teachers_qualified = teachers.all.to_a.count do |teacher|
+        UserLevel.where(user_id: teacher.id).passing.count > 10
+      end
+
+      # 50 * num teachers who have completed > 10 puzzles
+      payment_amount = 50 * num_teachers_qualified
```
Comment (mehalshah) [inline dashboard/app/models/pd/workshop_organizer_report.rb]: Define 50 as a constant somewhere

File: dashboard/app/models/pd/workshop_organizer_report.rb [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      Pd::Workshop::COURSE_CSP,
+      Pd::Workshop::COURSE_ECS,
+      Pd::Workshop::COURSE_CS_IN_A,
+      Pd::Workshop::COURSE_CS_IN_S
+    ].include?(workshop.course) && [
+      Pd::Workshop::TYPE_PUBLIC,
+      Pd::Workshop::TYPE_PRIVATE
+    ].include?(workshop.workshop_type)
+
+      qualified = teachers.count > 0
```
Comment (mehalshah) [inline dashboard/app/models/pd/workshop_organizer_report.rb]: Same with these constants - move as many of them as possible to constants or even a file that defines payment terms constants

File: dashboard/app/models/pd/workshop_organizer_report.rb [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      payment_amount = 40 * teachers.count
+
+      if plp && plp.urban?
+        payment_amount *= 1.25
+      end
+
+      payment_amount = (payment_amount + 500 * workshop.facilitators.count) * workshop.sessions.count
+    end
+
+    if workshop.course == Pd::Workshop::COURSE_COUNSELOR_ADMIN &&
```
Comment (mehalshah) [inline dashboard/app/models/pd/workshop_organizer_report.rb]: Make this an elsif

File: dashboard/app/models/pd/workshop_organizer_report.rb [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      subject: workshop.subject,
+      num_teachers: teachers.count,
+      days: workshop.sessions.count,
+      payment_type: payment_type,
+      qualified: qualified,
+      payment_amount: payment_amount
+    }
+  end
+
+  def self.calculate_payment(workshop, teachers, plp)
```
Comment (mehalshah) [inline dashboard/app/models/pd/workshop_organizer_report.rb]: Payment calculation belongs in its own class, and arguably multiple classes.  At the risk of nitpicking, it seems to me that reports should display results more than calculate those results in the first place.

Happy to talk more about it offline

File: dashboard/app/models/pd/teacher_progress_report.rb (hunk 3 above)
Comment (aoby) [inline dashboard/app/models/pd/teacher_progress_report.rb]: On second thought, this is only the model. The controller (forthcoming PR) will handle the permissions. I can't redirect at this level, so I'll just raise an error.

**Comment by aoby** [comment]
I updated the PR with the following changes:
1. Moved workshop organizer payment model into its own class. I clarified
   the various payment part names with PM and now expose each one so it's
   easier to follow, test, and reason about.
2. Renamed some of the strings for workshop course and subject on
   request from PM.
3. Incorporated earlier PR feedback.

File: dashboard/app/models/pd/session.rb [hunk 8]
```diff
@@ -18,6 +18,24 @@ class Pd::Session < ActiveRecord::Base
   belongs_to :workshop, class_name: 'Pd::Workshop', foreign_key: 'pd_workshop_id'
   has_many :attendances, class_name: 'Pd::Attendance', foreign_key: 'pd_session_id', dependent: :destroy
 
+  validates_presence_of :start, :end
+  validate :starts_and_ends_on_the_same_day
+  validate :starts_before_ends
+
+  def starts_and_ends_on_the_same_day
```
Comment (mehalshah) [inline dashboard/app/models/pd/session.rb:25]: While you are doing this you should have some testing for this

**Comment by mehalshah** [comment]
This looks fine but the payment calculation part is at the upper bound of desirable complexity. I think this is good to check in and revisit the calculators later

File: dashboard/app/models/pd/attendance.rb [hunk 9]
```diff
@@ -19,7 +19,19 @@ class Pd::Attendance < ActiveRecord::Base
 
   has_one :workshop, class_name: 'Pd::Workshop', through: :session
 
-  def self.for_teacher_in_workshop(teacher, workshop)
-    joins(:workshop).where(teacher: teacher, pd_workshops: {id: workshop.id})
+  def self.for_teacher(teacher)
+    self.joins(:workshop).where(teacher_id: teacher.id)
+  end
```
Comment (joshlory) [inline dashboard/app/models/pd/attendance.rb:24]: Nit: prefer [scopes](http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html#method-i-scope) over class methods here.  (Scopes are just syntactic sugar for the class method.)

File: dashboard/app/models/pd/session.rb (hunk 8 above)
Comment (aoby) [inline dashboard/app/models/pd/session.rb:25]: tests added

**Pull request merged**